libs/spock-shaded/src/main/java/org/codice/spock/extension/builtin/DeFinalizeClassLoader.java

```diff
@@ -0,0 +1,168 @@
+ * its parent. This classloader will therefore reload all classes while definalizing those that are
+ * requested except for all classes in the following packages:
+ *
+ * <ul>
+ *   <li>java
+ *   <li>javax
+ *   <li>sun
+ *   <li>org.xml
+ *   <li>org.junit
+ * </ul>
+ *
+ * These packages are not being reloaded as they are required by the {@link DeFinalizeSputnik} test
+ * runner to delegate to the real {@link org.spockframework.runtime.Sputnik} test runner.
+ *
+ * <p>All loaded classes will be done so with the same protection domain used if they had been
+ * loaded by the parent classloader.
+ */
+public class DeFinalizeClassLoader extends ClassLoader {
+  private final ClassPool pool;
+  private final Set<String> filters;
+
+  /**
+   * Constructs a new classloader for the specified Spock test specification class. The class is
+   * used to retrieve the {@link DeFinalize} annotations in order to identify which classes or
+   * packages should be definalized as they are being loaded.
+   *
+   * <p>The classloader used to load this classloader class will be defined as its parent
+   * classloader.
+   *
+   * @param specClass the Spock test specification class
+   */
+  public DeFinalizeClassLoader(Class<?> specClass) {
+    super(DeFinalizeClassLoader.class.getClassLoader());
+    this.pool = new ClassPool(false);
+    this.pool.appendClassPath(new LoaderClassPath(getParent()));
+    this.pool.appendSystemPath();
+    this.filters =
+        Stream.concat(
+                Stream.of(specClass.getAnnotationsByType(DeFinalize.class))
+                    .map(DeFinalize::value)
+                    .flatMap(Stream::of)
+                    .map(Class::getName),
+                Stream.of(specClass.getAnnotationsByType(DeFinalize.class))
+                    .map(DeFinalize::packages)
+                    .flatMap(Stream::of))
+            .collect(Collectors.toSet());
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Loads the class with the specified <a href="#name">binary name</a>.
+   *
+   * <p>The class will always be loaded first from its parent classloader. It will be returned as is
+   * if the class is from one of the reserved packages (see above). Otherwise, the class will be
+   * reloaded and optionally definalize if the class match a requested class or package from one of
+   * the {@link DeFinalize} annotations of the Spock test specification class. The protection domain
+   * from the version of the class loaded by the parent classloader will be re-used when defining
+   * the reloaded class.
+   *
+   * @param name the binary name of the class to load
+   * @param resolve <code>true</code> to resolve the class; <code>false</code> otherwise
+   * @return the resulting class object
+   * @throws ClassNotFoundException if the class could not be found
+   */
+  @Override
+  protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+    synchronized (getClassLoadingLock(name)) {
```

Comment:
❓ I'm wondering if you can deadlock here, in the case where there are circular or reciprocal references between classes. I'm reasonably certain you won't, but not 100%.

Reply:
That is exactly how the base class ends up synchronizing. In our case since this classloader is not parallel, the getClassLoadingLock() will return this and therefore no deadlock should occur